src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs

```diff
@@ -310,6 +313,34 @@ public override BoundNode VisitLabeledStatement(BoundLabeledStatement node)
         public override BoundNode VisitGotoStatement(BoundGotoStatement node)
         {
             _labelsUsed.Add(node.Label);
+
+            // check for illegal jumps across using declarations
+            var sourceStart = node.Syntax.Location.SourceSpan.Start;
+            var targetStart = node.Label.Locations[0].SourceSpan.Start;
+
+            foreach (var usingDecl in _usingDeclarations)
+            {
+                var usingStart = usingDecl.Locations[0].SourceSpan.Start;
+                if (sourceStart < usingStart && targetStart > usingStart)
+                {
+                    // No forward jumps
+                    Diagnostics.Add(ErrorCode.ERR_GoToForwardJumpOverUsingVar, node.Syntax.Location);
+                    break;
+                }
+                else if(sourceStart > usingStart && targetStart < usingStart)
+                {
+                    // Backwards jump, so we must have already seen the label
+                    Debug.Assert(_labelsDefined.ContainsKey(node.Label)); 
+
+                    // Error if label and using are part of the same block
+                    if (_labelsDefined[node.Label].Locals.Contains(usingDecl))
```

Comment:
>_labelsDefined[node.Label].Locals.Contains(usingDecl) [](start = 24, length = 53)

Perhaps use `PooledDictionary<LocalSymbol, BoundBlock> _usingDeclarations` so this check can be a simple comparison: `_labelsDefined[node.Label] == usingDecl.Value`.